src/Workspaces/Core/MSBuild/MSBuild/ProjectFile/ProjectFile.cs

```diff
@@ -56,32 +56,20 @@ public async Task<ImmutableArray<ProjectFileInfo>> GetProjectFileInfosAsync(Canc
                 // In this case, we need to iterate through the <TargetFrameworks>, set <TargetFramework> with
                 // each value, and build the project.
 
-                var hasTargetFrameworkProp = _loadedProject.GetProperty(PropertyNames.TargetFramework) != null;
                 var targetFrameworks = targetFrameworksValue.Split(';');
                 var results = ImmutableArray.CreateBuilder<ProjectFileInfo>(targetFrameworks.Length);
 
                 foreach (var targetFramework in targetFrameworks)
                 {
-                    _loadedProject.SetProperty(PropertyNames.TargetFramework, targetFramework);
+                    _loadedProject.SetGlobalProperty(PropertyNames.TargetFramework, targetFramework);
                     _loadedProject.ReevaluateIfNecessary();
 
                     var projectFileInfo = await BuildProjectFileInfoAsync(cancellationToken).ConfigureAwait(false);
 
                     results.Add(projectFileInfo);
                 }
 
-                // Remove the <TargetFramework> property if it didn't exist in the file before we set it.
-                // Otherwise, set it back to it's original value.
-                if (!hasTargetFrameworkProp)
-                {
-                    var targetFrameworkProp = _loadedProject.GetProperty(PropertyNames.TargetFramework);
-                    _loadedProject.RemoveProperty(targetFrameworkProp);
-                }
-                else
-                {
-                    _loadedProject.SetProperty(PropertyNames.TargetFramework, targetFrameworkValue);
-                }
-
+                _loadedProject.RemoveGlobalProperty(PropertyNames.TargetFramework);
```

Comment:
Removing this global property whether or not it had been set cannot result in any side effects?

Reply:
I thought it wasn't necessary with global properties, since the only case which could have been impacted is if someone instantiates a `MSBuildWorkspace` with a `TargetFramework` property set to an empty string, which seemed wrong in the first place.

But I guess someone could instantiate `MSBuildWorkspace` with an empty `TargetFramework` but also a non-empty `TargetFrameworks` property to force a single-targeted project to use a different set of TFMs.

I can add some defensive code to handle this case.

Reply:
That sounds like a nice change to me--I agree it's funky but possible.

Reply:
OK I added the safeguard, along with a simple unit test to document the possible use case.

I tried to find possible side effects of always removing the global property, and the only thing I could come up with is when all of the following conditions are combined:

 - You call `MSBuildWorkspace.TryApplyChanges`
 - On a `MSBuildWorkspace` with the `TargetFramework` global property set to an empty string
 - On a project that sets the `TargetFramework` property
 - With changes that affect a project which includes a condition on the `TargetFramework` property that would alter the effect of the MSBuild project `AddItem` or `RemoveItem` method (such as a conditional glob for instance)

Needless to say, that seems *very* contrived and unlikely, so I didn't try to write a unit test which would simulate such a mess, if you're OK with that. I don't think that would bring any real value.

Note that after loading the `MSBuildWorkspace` project, the MSBuild project is thrown away along with the `ProjectFile` instance. It's only used after the initial evaluation in `MSBuildWorkspace.TryApplyChanges`.